Features/absorption chillers

jakob-wo opened the pull request:
This PR introduces...

1. A module with (pre-)calculations for absorption chillers as described in Issue #69 
2. Examples on how to use the calculations
3. Documentation of the module

jakob-wo commented:
The checks fail because...

a) There is one test that does not pass yet. Test-driven-development! ;)

b) flake8 complains that there is a variable inside an if-else-loop that is assigned but not used. I do not completely understand the problem because I think the variable is used under certain conditions (triggering 'else').

> F841 local variable 'Q_dot_evap' is assigned to but never used

MaGering commented:
I would like to offer help reviewing, in case still needed. I am already familiar with absorption heat pumps (Type I and II) and the characteristic equation method.

jakob-wo commented:
> I would like to offer help reviewing

@MaGering Thank you - I would appreciate a review from you! I will let you know when it is ready to have a look. Right now I am still improving docstrings and docu... but in the next days it will be ready.

jakob-wo commented:
Hi @MaGering ,
docu and docstrings are finished now and the PR is ready for a review.
I am looking forward to feedback and ideas for improvments from you!

MaGering commented:
Wow! The calculation of the absorption chiller already works really well and is carefully elaborated! :)

I've started the review and will leave my comments directly in the files changed section.

MaGering commented on examples/absorption_heatpumps_and_chiller/data/characteristic_parameters.csv:
```diff
@@ -0,0 +1,6 @@
+Safarik,2,single-effect absorption chiller,15,2.138,2.252,0.452,-8.037,0.489,-4.913,,,
+Broad_01,3,single-effect absorption chiller,768,1.707,1.844,24.121,-553.194,24.121,-553.194,,,
+Broad_02,4,double-effect absorption chiller,1163,2.162,4.499,15.459,-911.572,10.807,-603.85,,,
+Kuehn,5,single-effect absorption chiller,10,2.5,1.8,0.42,0.9,0.51,2,,,
```
```suggestion
Kuehn,5,single-effect absorption chiller,10,2.5,1.8,0.42,0.9,0.53,1.72,,,
```

I'm wondering where the values 0.51 and 2 are coming from, since I haven't found them in "A 10 kW INDIRECTLY FIRED ABSORPTION HEAT PUMP: CONCEPTS FOR A REVERSIBLE OPERATION" [1]

This lead me to another general issue: For the calculation of the Generators / Desorbers Heat ∆∆t'' is needed instead of ∆∆t' as in the documentation mentioned. That's why I calculated in a separate script the values for s_G and r_G for ∆∆t' instead of ∆∆t''.

The calculation results in these two equations according to [1]:

Q_D = 0.53 * ∆∆t'' - 2.5                   (I)
Q_D = 0.53 * ∆∆t' + 1.716667         (II)

If I understood correctly we want to use equation II. Is that correct?

jakob-wo replied:
I am sorry for the confusion I created!
You are right, the values 0.51 and 2 are not in [1] but coming from another paper. I mixed them both up. The correct paper instead would be: 
_Kühn A., F. Ziegler 2005. “Operational results of a 10 kW absorption chiller and adaptation of the characteristic equation”, Proc. of the 1st Int. Conf. Solar Air Conditioning, 6-7 October 2005, Bad Staffelstein, Germany._ [3]
[3] and [1] both deal with the same machine but illustrate different approaches (improvements) of the characteristic equation.
The values 0.51 and 2 then are ment to be used with ∆∆t', like in (II) as you already figuered out correctly.

MaGering replied:
Alright! :) Do we need [1] then or can we replace it with [3] in the documentation?

MaGering commented on examples/absorption_heatpumps_and_chiller/absorption_chiller.py:
```diff
@@ -0,0 +1,141 @@
+solver = 'cbc'
+debug = False
+number_of_time_steps = 48
+periods = number_of_time_steps
```
Isn't 'periods' redundant?

jakob-wo replied:
yes it is. I'll remove it.

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+points of operation are given.
+You find detailed information in the referenced papers.
+
+This package comes with characteristic parameters for five absorption chillers
+published by Puig-Arnavat et al. [2]:
+'Rotartica', 'Safarik', 'Broad_01', 'Broad_02', 'Kuehn'.
```
```suggestion
This package comes with characteristic parameters for five absorption chillers.
Four published by Puig-Arnavat et al. [2]: 'Rotartica', 'Safarik', 'Broad_01' and 'Broad_02' 
and one published by Kühn and Ziegler [1]: 'Kuehn'.
```

jakob-wo replied:
Thank you! :)

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+
+The cooling capacity (:math:`\dot{Q}_{E}`) is determined by a function of
+the characteristic
+temperature difference (:math:`\Delta\Delta T`) that combines the external
```
```suggestion
temperature difference (:math:`\Delta\Delta t`) that combines the external
```

I would rather replace uppercase letter 'T' with lowercase letter 't' since the uppercase letter stands for internal temperatures and might be misleading. The replacement would also be consistent with the literature we use. It concerns the formulas in the docs and the docs schematic.

jakob-wo replied:
You are absolutly right. I'll adapt it.

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+absorption chillers.
+
+
+.. figure:: _pics/abs_chiller_col_02.png
```
Is it possible to add a dot to heat flows in the schematic?

jakob-wo replied:
Done!

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+
+    COPs = Q_dots_evap / Q_dots_gen
+
+Fig.2 illustrates how the cooling capacity and the COP of an absorption
```
```suggestion
Fig. 2 illustrates how the cooling capacity and the COP of an absorption
```

jakob-wo replied:
I prefere to stick with _Fig.2_ or rename it _Figure 2_.
But I realized the other modules do not have captions for their figueres... Personally, I am a fan of captions but we should try to unify the appearance...

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+    COPs = Q_dots_evap / Q_dots_gen
+
+Fig.2 illustrates how the cooling capacity and the COP of an absorption
+chiller (here the 10kW absorption chiller mentioned above) depend on the
```
```suggestion
chiller (here the 10 kW absorption chiller mentioned above) depend on the
```

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+    :align: center
+    :figclass: align-center
+
+    Fig.2: Dependency of the cooling capacity and the COP of a 10kW absorption
+    chiller on the cooling water temperature.
+
+You find the code that is behind Fig.2 in our examples:
```
```suggestion
    Fig. 2: Dependency of the cooling capacity and the COP of a 10 kW absorption
    chiller on the cooling water temperature.

You find the code that is behind Fig. 2 in our examples:
```

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+To model one of the machines provided by this package you can adapt the code
+above in the following way. You find information on the machines in [2].
+
+.. code-block:: python
```
I think this code-block needs to be updated? It has changed in the example.

jakob-wo replied:
Sorry, I don't see it... what has changed? What do I need to update exactly?

MaGering replied:
Sorry for that unhandy comment. The code is not running due to just a few things:

- pandas and os imports are missing
- The parameter 'R_G' needs to be changed to 'r_G'
- Q_dots_evap / Q_dots_gen is not possible for list type

You can find my suggestion below.

MaGering commented on examples/absorption_heatpumps_and_chiller/absorption_chiller.py:
```diff
@@ -0,0 +1,141 @@
+    coef_r=charpara[(charpara['name'] == chiller_name)]['r_G'].values[0],
+    method='kuehn_and_ziegler')
+COPs = [Qevap / Qgen for Qgen, Qevap in zip(Q_dots_gen, Q_dots_evap)]
+nominal_value = 25
```
How has this value been chosen?

jakob-wo replied:
Because it is operating as chiller it should be the evaporator heat Q_E and therefore 10 kW, right?
I guess I used random values when I first set up this example and finally forgot to adjust them :)

MaGering replied:
I see. I was confused with the nominal value because I didn't know which one was meant. I compared this example to the 'airsource_heatpump_const_max_output.py' example. There, the nominal values are called nominal conditions and contain 'nominal_Q_hot' and 'nominal_el_consumption', which makes it more clear in my opinion. That's why I would suggest to change 'nominal_value' to 'nominal_Q_dots_evap' in this example.

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,215 @@
+
+Fig.2 illustrates how the cooling capacity and the COP of an absorption
+chiller (here the 10kW absorption chiller mentioned above) depend on the
+colling water temperature, i.e. the mean external fluid temperature at
```
```suggestion
cooling water temperature, i.e. the mean external fluid temperature at
```

MaGering commented on examples/absorption_heatpumps_and_chiller/cooling_cap_dependence_on_cooling_water_temp.py:
```diff
@@ -0,0 +1,86 @@
+    method='kuehn_and_ziegler')
+
+
+fig1 = plt.figure()
```
```suggestion
fig1 = plt.figure(figsize=(8, 6))
```
If plt.show() is not commented out the figure is cut off in parts. Setting figsize solves this problem.

MaGering commented on src/oemof/thermal/absorption_heatpumps_and_chillers.py:
```diff
@@ -0,0 +1,202 @@
+    return Q_dots
+
+
+def define_AC_specs(Q_dots_evap, Q_dots_gen):
```
Is there already an example where this function is used?

jakob-wo replied:
No, not yet.

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,212 @@
+
+.. code-block:: python
+
+    import oemof.thermal.absorption_heatpumps_and_chillers as abs_chiller
+
+    filename_charpara = os.path.join(os.path.dirname(__file__), 'data/characteristic_parameters.csv')
+    charpara = pd.read_csv(filename_charpara)
+
+    chiller_name = 'Kuehn'  # 'Rotartica', 'Safarik', 'Broad_01', 'Broad_02'
+
+     # Characteristic temperature difference
+    ddt = abs_chiller.calc_characteristic_temp(
+        t_hot=[85],  # in °C
+        t_cool=[26],  # in °C
+        t_chill=[15],  # in °C
+        coef_a=charpara[(charpara['name'] == chiller_name)]['a'].values[0],
+        coef_e=charpara[(charpara['name'] == chiller_name)]['e'].values[0],
+        method='kuehn_and_ziegler')
+
+    # Cooling capacity
+    Q_dots_evap = abs_chiller.calc_heat_flux(
+        ddts=ddt,
+        coef_s=charpara[(charpara['name'] == chiller_name)]['s_E'].values[0],
+        coef_r=charpara[(charpara['name'] == chiller_name)]['r_E'].values[0],
+        method='kuehn_and_ziegler')
+
+    # Driving heat
+    Q_dots_gen = abs_chiller.calc_heat_flux(
+        ddts=ddt,
+        coef_s=charpara[(charpara['name'] == chiller_name)]['s_G'].values[0],
+        coef_r=charpara[(charpara['name'] == chiller_name)]['R_G'].values[0],
+        method='kuehn_and_ziegler')
+
+    COPs = Q_dots_evap / Q_dots_gen
```
```suggestion
    import oemof.thermal.absorption_heatpumps_and_chillers as abs_chiller
    import pandas as pd
    import os

    filename_charpara = os.path.join(os.path.dirname(__file__), 'data/characteristic_parameters.csv')
    charpara = pd.read_csv(filename_charpara)

    chiller_name = 'Kuehn'  # 'Rotartica', 'Safarik', 'Broad_01', 'Broad_02'

     # Characteristic temperature difference
    ddt = abs_chiller.calc_characteristic_temp(
        t_hot=[85],  # in °C
        t_cool=[26],  # in °C
        t_chill=[15],  # in °C
        coef_a=charpara[(charpara['name'] == chiller_name)]['a'].values[0],
        coef_e=charpara[(charpara['name'] == chiller_name)]['e'].values[0],
        method='kuehn_and_ziegler')

    # Cooling capacity
    Q_dots_evap = abs_chiller.calc_heat_flux(
        ddts=ddt,
        coef_s=charpara[(charpara['name'] == chiller_name)]['s_E'].values[0],
        coef_r=charpara[(charpara['name'] == chiller_name)]['r_E'].values[0],
        method='kuehn_and_ziegler')

    # Driving heat
    Q_dots_gen = abs_chiller.calc_heat_flux(
        ddts=ddt,
        coef_s=charpara[(charpara['name'] == chiller_name)]['s_G'].values[0],
        coef_r=charpara[(charpara['name'] == chiller_name)]['r_G'].values[0],
        method='kuehn_and_ziegler')

    COPs = [Qevap / Qgen for Qgen, Qevap in zip(Q_dots_gen, Q_dots_evap)]
```

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,212 @@
+        coef_r=2,
+        method='kuehn_and_ziegler')
+
+    COPs = Q_dots_evap / Q_dots_gen
```
```suggestion
    COPs = [Qevap / Qgen for Qgen, Qevap in zip(Q_dots_gen, Q_dots_evap)]
```

MaGering commented on examples/absorption_heatpumps_and_chiller/absorption_chiller.py:
```diff
@@ -0,0 +1,140 @@
+    coef_s=charpara[(charpara['name'] == chiller_name)]['s_G'].values[0],
+    coef_r=charpara[(charpara['name'] == chiller_name)]['r_G'].values[0],
+    method='kuehn_and_ziegler')
+COPs = [Qevap / Qgen for Qgen, Qevap in zip(Q_dots_gen, Q_dots_evap)]
+nominal_value = 25
+actual_value = [Q_e / nominal_value for Q_e in Q_dots_evap]
+actual_value_df = pd.DataFrame(actual_value).clip(0)
+
+# Absorption Chiller
+energysystem.add(solph.Transformer(
+    label="AC",
+    inputs={b_th_high: solph.Flow()},
+    outputs={b_th_low: solph.Flow(nominal_value=nominal_value,
```
```suggestion
COPs = [Qevap / Qgen for Qgen, Qevap in zip(Q_dots_gen, Q_dots_evap)]
nominal_Q_dots_evap = 10
actual_value = [Q_e / nominal_Q_dots_evap for Q_e in Q_dots_evap]
actual_value_df = pd.DataFrame(actual_value).clip(0)

# Absorption Chiller
energysystem.add(solph.Transformer(
    label="AC",
    inputs={b_th_high: solph.Flow()},
    outputs={b_th_low: solph.Flow(nominal_value=nominal_Q_dots_evap,
```

jakob-wo commented:
@MaGering Thank you so much!
I appreaciate your work and effort - you're helping me a lot.

jakob-wo commented:
I updated the docu.
@MaGering could you have another look at it? 

I'd like to keep all three references in the docu, even though [2] (former [1]) is not necessary any more but it is open access whereas [1] is not. What do you think?
To prevent confusions about the different parameters in the references I added an aditional paragraph in the end of the docu.

jakob-wo commented:
Now, _coverall_ complains because testcoverage decreased...
Well, I think that is aceptable :)

jnnr commented:
> Now, _coverall_ complains because testcoverage decreased...
> Well, I think that is aceptable :)

You can have a look in the coverage reports to see which lines of the ones you introduced are not tested (https://coveralls.io/builds/31085093/source?filename=src%2Foemof%2Fthermal%2Fabsorption_heatpumps_and_chillers.py#L76)

jnnr commented on src/oemof/thermal/absorption_heatpumps_and_chillers.py:
```diff
@@ -0,0 +1,207 @@
+        list_t_ac = t_cool
+    else:
+        print("")
+        print("ERROR - "
```
Printing to inform users of errors is not common. Instead, raise an error (as you do some lines above).

MaGering commented on docs/absorption_chillers.rst:
```diff
@@ -0,0 +1,220 @@
+You find information on the machines in [1], [2] and [3].
+Please be aware that [2] introduces a slightly different approach
+(using an improved characteristic equation with :math:`\Delta\Delta t''`
+instead of :math:`\Delta\Delta t'`)
+and therefore find different characteristic parameters that the ones from [1] we
+are using here.
```
```suggestion
instead of :math:`\Delta\Delta t'`).
The characteristic parameters that we use are derived from [2] and therefore differ from those in [1].
```

MaGering commented:
> I'd like to keep all three references in the docu, even though [2] (former [1]) is not necessary any more but it is open access whereas [1] is not. What do you think?
> To prevent confusions about the different parameters in the references I added an aditional paragraph in the end of the docu.

I like that solution! With the extra paragraph it is clear where the characteristic parameter come from. I adjusted the wording slightly.

jakob-wo commented:
@jnnr: The Merging is blocked because you requested changes. I fixed the error handling you mentioned but I do not see what else you requested - What else do I have to adapt to fulfill the requested changes? Would you give me a hint, please :)

jnnr commented:
> @jnnr: The Merging is blocked because you requested changes. I fixed the error handling you mentioned but I do not see what else you requested - What else do I have to adapt to fulfill the requested changes? Would you give me a hint, please :)

I had a quick look and saw that now, errors are raised. Thanks for fixing this! I did not have a full review though and I did not test the code. If you need another review, I could do one, but not within the next two weeks.
